### PR TITLE
Use PODArray to render in sparkBar function

### LINKED ISSUE #44467: sparkbar aggregate function causes server to be killed with OOM
**Describe what's wrong**
Using `sparkbar` aggregate function on tables that contain large UInt64 values causes OOM (Out of Memory) Killer process to kill ClickHouse server irrespective of any memory usage restrictions set.

**Does it reproduce on recent release?**
Yes. Any version `>=21.11` including the latest `22.12`. Also reproducible on ClickHouse.Cloud.

**How to reproduce**

```sql
DROP TABLE IF EXISTS test; 
CREATE TABLE test (x UInt64, y UInt8) Engine=MergeTree ORDER BY tuple();
INSERT INTO test VALUES (18446744073709551615,255),(0,0),(0,0),(4036797895307271799,163)
SELECT sparkbar(9)(x,y) FROM test;
```

**Expected behavior**

Memory restrictions should be enforced and server should not be killed with OOM.

**Error message and/or stacktrace**

```
2022.12.20 01:48:30.983421 [ 1 ] {} <Fatal> Application: Child process was terminated by signal 9 (KILL). If it is not done by 'forcestop' command or manually, the possible cause is OOM Killer (see 'dmesg' and look at the '/var/log/kern.log' for the details).
```

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Use PODArray to render in sparkBar function, so we can control the memory usage. Close #44467

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by nickitat — APPROVED** [review]

**Comment by nickitat** [comment]
Stateless tests (release, DatabaseReplicated) [4/4] - #44542

File: src/AggregateFunctions/AggregateFunctionSparkbar.h [hunk 1]
```diff
@@ -242,11 +248,13 @@ class AggregateFunctionSparkbar final
 
             auto get_bars = [&] (const std::optional<Float64> & point_y)
             {
-                value += getBar(point_y ? std::round(((point_y.value() - min_y.value()) / diff_y) * 7) + 1 : 0);
+                auto bar = getBar(point_y ? std::round(((point_y.value() - min_y.value()) / diff_y) * 7) + 1 : 0);
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/AggregateFunctionSparkbar.h:251]: The function getBar can insert directly into the destination array.

**Comment by alexey-milovidov** [comment]
I've edited and replaced PaddedPODArray to PODArray because I cannot figure out why padding has any use here.
I've removed an unneeded header.

Also, I don't understand why MemoryTracker does not work from the `operator new`, which should be inside the String.
@nickitat @azat what do you think?

File: src/AggregateFunctions/AggregateFunctionSparkbar.h [hunk 2]
```diff
@@ -319,7 +327,7 @@ class AggregateFunctionSparkbar final
     {
         auto & to_column = assert_cast<ColumnString &>(to);
         const auto & data = this->data(place);
-        const String & value = render(data);
+        const auto & value = render(data);
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/AggregateFunctionSparkbar.h:330]: Let's make the `render` function insert directly into the column, and avoid any copying, and then even a `using` directive will be unneeded.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Comment by azat** [comment]
> Also, I don't understand why MemoryTracker does not work from the operator new, which should be inside the String.

Me neither, but this has been changed in #24483.

Well, I do understand that this can improve exception safety, but not sure that it is the problem right now - https://github.com/ClickHouse/ClickHouse/issues/24110#issuecomment-841616046

**Comment by canhld94** [comment]
Perfect green donut... 
@nickitat @alexey-milovidov can I get some reviews [THANKS]

**Comment by nickitat** [comment]
> Perfect green donut...
> @nickitat @alexey-milovidov can I get some reviews [THANKS]

looks like Alexey's suggestion about `render` isn't fixed yet

File: src/AggregateFunctions/AggregateFunctionSparkbar.h [hunk 3]
```diff
@@ -104,14 +105,14 @@ class AggregateFunctionSparkbar final
...
-    using Chars = PODArray<char>;
 
     template <class T>
-    void updateFrame(Chars & frame, const T value) const
+    int updateFrame(ColumnString::Chars & frame, const T value) const
     {
         static const String bars[9] = {" ", "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"};
         const auto & bar = (isNaN(value) || value > 8 || value < 1) ? bars[0] : bars[static_cast<UInt8>(value)];
         frame.insert(bar.begin(), bar.end());
+        return static_cast<int>(bar.size());
```
Comment (nickitat) [inline src/AggregateFunctions/AggregateFunctionSparkbar.h:115]: why it should be an int?

Comment (canhld94) [reply src/AggregateFunctions/AggregateFunctionSparkbar.h:115]: `bar` values are limited to 9 values, which is quite small, so I think using `int` is enough.

File: src/AggregateFunctions/AggregateFunctionSparkbar.h [hunk 4]
```diff
@@ -105,24 +107,12 @@ class AggregateFunctionSparkbar final
     bool specified_min_max_x;
 
     template <class T>
-    String getBar(const T value) const
+    int updateFrame(ColumnString::Chars & frame, const T value) const
```
Comment (vdimir) [inline src/AggregateFunctions/AggregateFunctionSparkbar.h:110]: `updateFrame` is more complicated than `getBar`, what is the key difference?

File: src/AggregateFunctions/AggregateFunctionSparkbar.h (hunk 3 above)
Comment (alexey-milovidov) [reply src/AggregateFunctions/AggregateFunctionSparkbar.h:115]: But it also cannot be negative - so the type should be unsigned.

File: src/AggregateFunctions/AggregateFunctionSparkbar.h (hunk 4 above)
Comment (canhld94) [reply src/AggregateFunctions/AggregateFunctionSparkbar.h:110]: Before we get a string (which create a copy), insert it into the the column (which is another copy). `updateFrame` will keep a static list of strings, then select the proper value from the list, then insert it into the column.

**Review by vdimir — APPROVED** [review]

**Comment by vdimir** [comment]
@alexey-milovidov could you please approve (or comment)? Merging is blocked

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**